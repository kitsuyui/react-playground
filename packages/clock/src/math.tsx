import { DateTime } from 'luxon'

import type { StepStyle, HMS } from './types'

export function hourToDegree(hour: number): number {
  return hour * 30
}

export function secondToDegree(second: number): number {
  return second * 6
}

export function minuteToDegree(minute: number): number {
  return minute * 6
}

export function degreeToSin(degree: number): number {
  return Math.sin(degreeToRadian(degree))
}

export function degreeToCos(degree: number): number {
  return Math.cos(degreeToRadian(degree))
}

export function degreeToRadian(degree: number): number {
  return degree * (Math.PI / 180)
}

export function calcHMS(
  date: Date,
  timezone: string,
  stepStyle: StepStyle
): HMS {
  switch (stepStyle) {
    case 'tick':
      return tickHMS(date, timezone)
    case 'sweep':
      return sweepHMS(date, timezone)
  }
}

export function tickHMS(date: Date, timezone: string): HMS {
  const datetime = DateTime.fromJSDate(date)
  const dt = datetime.setZone(timezone)
  const hour = dt.hour
  const minute = dt.minute
  const second = dt.second
  return { hour, minute, second }
}

export function sweepHMS(date: Date, timezone: string): HMS {
  const datetime = DateTime.fromJSDate(date)
  const dt = datetime.setZone(timezone)
  const hour = dt.hour + dt.minute / 60 + dt.second / 3600
  const minute = dt.minute + dt.second / 60
  const second = dt.second + dt.millisecond / 1000
  return { hour, minute, second }
}
