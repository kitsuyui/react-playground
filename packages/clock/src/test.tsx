import { render } from '@testing-library/react'
import React from 'react'

import { tickHMS, sweepHMS, calcHMS } from './math'

import {
  AnalogClock,
  DigitalClock,
  computeFace,
  romanFace,
  arabicFace,
} from '.'

test('render AnalogClock', () => {
  const dt = new Date('2021-01-01T00:00:00Z')
  const { asFragment } = render(<AnalogClock timezone="Asia/Tokyo" date={dt} />)
  expect(asFragment()).toMatchSnapshot()
})

test('render DigitalClock', () => {
  const dt = new Date('2021-01-01T00:00:00Z')
  const { asFragment } = render(
    <DigitalClock timezone="Asia/Tokyo" date={dt} />
  )
  expect(asFragment()).toMatchSnapshot()
})

test('tickHMS returns HMS', () => {
  const dt = new Date('2021-01-01T00:00:00Z')
  expect(tickHMS(dt, 'Asia/Tokyo')).toEqual({
    hour: 9,
    minute: 0,
    second: 0,
  })

  const dt2 = new Date('2021-01-01T10:08:14Z')
  expect(tickHMS(dt2, 'Asia/Tokyo')).toEqual({
    hour: 19,
    minute: 8,
    second: 14,
  })
})

test('sweepHMS returns HMS', () => {
  const dt2 = new Date('2021-01-01T10:08:14.2000Z')
  expect(sweepHMS(dt2, 'Asia/Tokyo')).toEqual({
    hour: 19.13722222222222,
    minute: 8.233333333333333,
    second: 14.2,
  })
})

test('calcHMS returns HMS', () => {
  const dt = new Date('2021-01-01T10:08:14.2000Z')
  expect(calcHMS(dt, 'Asia/Tokyo', 'tick')).toEqual(tickHMS(dt, 'Asia/Tokyo'))
  expect(calcHMS(dt, 'Asia/Tokyo', 'sweep')).toEqual(sweepHMS(dt, 'Asia/Tokyo'))
})

test('arabicFace returns arabicFace', () => {
  expect(arabicFace(0)).toBe('12')
  expect(arabicFace(1)).toBe('1')
  expect(arabicFace(2)).toBe('2')
  expect(arabicFace(3)).toBe('3')
  expect(arabicFace(4)).toBe('4')
  expect(arabicFace(5)).toBe('5')
  expect(arabicFace(6)).toBe('6')
  expect(arabicFace(7)).toBe('7')
  expect(arabicFace(8)).toBe('8')
  expect(arabicFace(9)).toBe('9')
  expect(arabicFace(10)).toBe('10')
  expect(arabicFace(11)).toBe('11')
})

test('romanFace returns romanFace', () => {
  expect(romanFace(0)).toBe('Ⅻ')
  expect(romanFace(1)).toBe('Ⅰ')
  expect(romanFace(2)).toBe('Ⅱ')
  expect(romanFace(3)).toBe('Ⅲ')
  expect(romanFace(4)).toBe('Ⅳ')
  expect(romanFace(5)).toBe('Ⅴ')
  expect(romanFace(6)).toBe('Ⅵ')
  expect(romanFace(7)).toBe('Ⅶ')
  expect(romanFace(8)).toBe('Ⅷ')
  expect(romanFace(9)).toBe('Ⅸ')
  expect(romanFace(10)).toBe('Ⅹ')
  expect(romanFace(11)).toBe('Ⅺ')
})

test('computeFace returns face', () => {
  expect(computeFace(0, 'roman')).toBe(romanFace(0))
  expect(computeFace(0, 'arabic')).toBe(arabicFace(0))
})
