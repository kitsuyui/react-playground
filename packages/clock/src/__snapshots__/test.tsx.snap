// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render AnalogClock 1`] = `
<DocumentFragment>
  <div
    style="width: 100px; height: 100px;"
  >
    <svg
      fill="none"
      height="100"
      viewBox="0 0 100 100"
      width="100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="50"
        cy="50"
        fill="white"
        r="45"
        stroke="black"
        stroke-width="2"
      />
      <circle
        cx="50"
        cy="50"
        fill="black"
        r="2"
        stroke="black"
        stroke-width="2"
      />
      <line
        stroke="black"
        stroke-width="2"
        x1="50"
        x2="20"
        y1="50"
        y2="50.00000000000001"
      />
      <line
        stroke="black"
        stroke-width="2"
        x1="50"
        x2="50"
        y1="50"
        y2="10"
      />
      <line
        stroke="red"
        stroke-width="2"
        x1="50"
        x2="50"
        y1="50"
        y2="10"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`render DigitalClock 1`] = `
<DocumentFragment>
  <div
    style="width: 20em; font-family: monospace;"
  >
    09:00:00 AM (Asia/Tokyo)
  </div>
</DocumentFragment>
`;
